geps/gep-1762.md

```diff
@@ -0,0 +1,294 @@
+exclude these from irrelevant resources.
+
+This is intended to support integration with the kitchen-sync of Kubernetes extensions which rely on labels and annotations,
+such as as prometheus scraping, object grouping/organization, etc.
+
+#### Arbitrary Customization
+
+GEP-1867 introduces a new `infrastructure` field, which allows customization of some common configurations (version, size, etc)
+and allows a per-Gateway generic `parametersRef`.
+This can be utilized for the remainder of customizations.
+
+### Resource Attachment
+
+Resources Generated in response to the `Gateway` are guaranteed to have two attributes:
+* A `gateway.networking.k8s.io/metadata.name: <NAME>` label.
+* A name `<NAME>-<GATEWAY CLASS>`.
+
+The generated resources MUST be in the same namespaces as the `Gateway`.
+
+Implementations MAY set `ownerReferences` to the `Gateway` in most cases, as well, but this is not required
+as some implementations may have different cleanup mechanisms.
+
+
+These can be relied on to attach resources to.
+While "Policy attachment" in the gateway-api would use attachment to the actual `Gateway` resource itself,
+many existing things attach only to resources like `Deployment` or `Service`.
+
+An example using these:
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: gateway
+spec:
+  gatewayClassName: example
+  listeners:
+  - name: default
+    hostname: "example.com"
+    port: 80
+    protocol: HTTP
+---
+apiVersion: autoscaling/v2
+kind: HorizontalPodAutoscaler
+metadata:
+  name: gateway
+spec:
+  # Match the generated Deployment by reference
+  # Note: Do not use `kind: Gateway`.
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: gateway-example
+  minReplicas: 2
+  maxReplicas: 5
+  metrics:
+  - type: Resource
+    resource:
+      name: cpu
+      target:
+        type: Utilization
+        averageUtilization: 50
+---
+apiVersion: policy/v1
+kind: PodDisruptionBudget
+metadata:
+  name: gateway
+spec:
+  minAvailable: 1
+  selector:
+    # Match the generated Deployment by label
+    matchLabels:
+      gateway.networking.k8s.io/metadata.name: gateway
+```
+
+Note: there is [discussion](https://github.com/kubernetes-sigs/gateway-api/discussions/1355) around a way to attach a HPA to a Gateway directly.
+
+## API
+
+This GEP extends the `infrastructure` API introduced in [GEP-1867](https://gateway-api.sigs.k8s.io/geps/gep-1867).
+
+```go
+type GatewayInfrastructure struct {
+	// Labels that should be applied to any resources created in response to this Gateway.
+	//
+	// For implementations creating other Kubernetes objects, this should be the `metadata.labels` field on resources.
+	// For other implementations, this refers to any relevant (implementation specific) "labels" concepts.
+	//
+  // If Labels is set on the GatewayClass as well, the labels are merged with the Gateway's labels taking precedence.
+  //
```

Comment:
Can we just start with this on the Gateway and consider support on `GatewayClass` as future work? I think we could make a good argument for precedence to go in either direction here and/or to recreate the `overrides` and `defaults` concept from inherited policy attachment in the `GatewayClass.spec.infrastructure` field to enable both. I think that's a sufficiently complex conversation that it can be saved for a follow up and just removed from the initial scope of this GEP.

Reply:
I had it like that at first and @youngnick suggesting it would be a good requirement. I don't have a preference either way so will do whatever

Reply:
@youngnick's traveling for a bit so may take some time to respond, but I did get some time to talk with him about this yesterday and he was on board with only targeting Gateway to start.

Reply:
Yes, this is on me, I think that having this work for Gateway first is better. Let's get something first, and add the GatewayClass later if needed.